### PR TITLE
fix: ESLToggleable fix focus flow for nested cases. Introducing focus manager

### PR DESCRIPTION
**Pull request by ala-n**

### CODE REVIEW

**Review by abarmina — APPROVED** [review]

**Review by NastaLeo — APPROVED** [review]

**Review by yadamskaya — APPROVED** [review]

File: src/modules/esl-utils/dom/focus.ts [hunk 1]
```diff
@@ -19,29 +19,11 @@ export const handleFocusChain = (e: KeyboardEvent, first: HTMLElement | undefine
...
-  if (type === 'loop') return handleFocusChain(e, $first, $last);
-
-  if (type === 'chain' && $fallback) {
-    if ($last && e.target !== (e.shiftKey ? $first : $last)) return;
-    $fallback.focus();
-    e.preventDefault();
-  }
-};
-
 const FOCUSABLE_SELECTOR = 'a[href], button, input, textarea, select, details, summary, output, [tabindex]:not([tabindex="-1"])';
```
Comment (dshovchko) [inline src/modules/esl-utils/dom/focus.ts:22]: Pls, consider improving the selector to exclude elements that have attribute "tabindex" with negative values.

File: src/modules/esl-toggleable/core/esl-toggleable-focus.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  /** Change focus scope to the specified element. Previous scope saved in the stack. */
+  public attach(element: ESLToggleable): void {
+    if (element.focusBehavior === 'none' && element !== this.current) return;
+    // Remove the element from the stack and add it on top
+    this.stack = this.stack.filter((el) => el !== element).concat(element);
+    // Focus on the first focusable element
+    queueMicrotask(() => afterNextRender(() => element.focus({preventScroll: true})));
+  }
+
+  /** Remove the specified element from the known focus scopes. */
```
Comment (dshovchko) [inline src/modules/esl-toggleable/core/esl-toggleable-focus.ts:45]:
```suggestion
  /** Removes the specified element from the known focus scopes. */
```

File: src/modules/esl-toggleable/core/esl-toggleable-focus.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  public get current(): ESLToggleable {
+    return this.stack[this.stack.length - 1];
+  }
+
+  /** Check if the element is in the known focus scopes */
+  public has(element: ESLToggleable): boolean {
+    return this.stack.includes(element);
+  }
+
+  /** Change focus scope to the specified element. Previous scope saved in the stack. */
```
Comment (dshovchko) [inline src/modules/esl-toggleable/core/esl-toggleable-focus.ts:36]:
```suggestion
  /** Changes focus scope to the specified element. Previous scope saved in the stack. */
```

File: src/modules/esl-toggleable/core/esl-toggleable-focus.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    instance = this;
+  }
+
+  /** Current focus scope */
+  public get current(): ESLToggleable {
+    return this.stack[this.stack.length - 1];
+  }
+
+  /** Check if the element is in the known focus scopes */
```
Comment (dshovchko) [inline src/modules/esl-toggleable/core/esl-toggleable-focus.ts:31]:
```suggestion
  /** Checks if the element is in the known focus scopes */
```

**Review by dshovchko — APPROVED** [review]

**Pull request merged**